subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/IvyComponentParser.java

```diff
@@ -177,50 +171,48 @@ private static String artifactKey(PublishArtifact publishArtifact) {
 
+        PublicationWarningsCollector publicationWarningsCollector =
+            new PublicationWarningsCollector(LOG, UNSUPPORTED_FEATURE, "", PUBLICATION_WARNING_FOOTER, "suppressIvyMetadataWarningsFor");
+
         for (SoftwareComponentVariant variant : component.getUsages()) {
-            publicationWarningsCollector.newContext(variant.getName());
+            VariantWarningCollector warnings = publicationWarningsCollector.warningCollectorFor(variant.getName());
 
-            ImmutableAttributes attributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
-            VariantVersionMappingStrategyInternal variantVersionMappingStrategy = versionMappingStrategy.findStrategyForVariant(attributes);
-            VariantDependencyFactory dependencyFactory = new VariantDependencyFactory(projectDependencyResolver, variantVersionMappingStrategy, publicationWarningsCollector);
+            VariantDependencyResolver dependencyResolver = variantDependencyResolverFactory.createResolver(variant, (organization, module, declaredVersion) -> {
+                if (declaredVersion == null) {
+                    // Version mapping is disabled or did not discover coordinates.
+                    warnings.addUnsupported(String.format("%s:%s declared without version", organization, module));
+                }
+                return declaredVersion;
+            });
+
+            VariantDependencyFactory dependencyFactory = new VariantDependencyFactory(dependencyResolver, warnings);
 
             for (ModuleDependency dependency : variant.getDependencies()) {
                 String confMapping = confMappingFor(variant, dependency);
-                if (!dependency.getAttributes().isEmpty()) {
-                    publicationWarningsCollector.addUnsupported(String.format("%s:%s:%s declared with Gradle attributes", dependency.getGroup(), dependency.getName(), dependency.getVersion()));
-                }
-                if (dependency instanceof ProjectDependency) {
-                    ivyDependencies.add(dependencyFactory.asProjectDependency((ProjectDependency) dependency, confMapping));
-                } else {
-                    ExternalDependency externalDependency = (ExternalDependency) dependency;
-                    if (platformSupport.isTargetingPlatform(dependency)) {
-                        publicationWarningsCollector.addUnsupported(String.format("%s:%s:%s declared as platform", dependency.getGroup(), dependency.getName(), dependency.getVersion()));
-                    }
-                    ivyDependencies.add(dependencyFactory.asExternalDependency(externalDependency, confMapping));
+                if (platformSupport.isTargetingPlatform(dependency)) {
+                    warnings.addUnsupported(String.format("%s:%s:%s declared as platform", dependency.getGroup(), dependency.getName(), dependency.getVersion()));
```

Comment:
What makes you say this is better? 

I'm not necessarily against it, though that means for each different kind of warning we'd need to add new methods to `VariantWarningCollector`. Plus, each metadata type will likely have different types of warnings, meaning most methods are likely to only be used once

Reply:
It keeps the interactions with the warning collector at a higher and more consistent level of abstraction rather than putting the details of string formatting right there all the time.

Reply:
I agree the `VariantWarningCollector` could use some work, though this was existing code so I would prefer to leave it as-is for now. 
As we implement POM rewriting I know we will have more warnings to emit, so I know this will be revisited. 

We should also think about migrating these warnings to use the Problems API